source/util/bit_stream.cpp

```diff
@@ -0,0 +1,379 @@
+                                             size_t chunk_length) {
+  return ReadVariableWidthUnsigned(this, val, chunk_length);
+}
+
+bool BitReaderInterface::ReadVariableWidthS64(int64_t* val,
+                                              size_t chunk_length,
+                                              size_t zigzag_exponent) {
+  return ReadVariableWidthSigned(this, val, chunk_length, zigzag_exponent);
+}
+
+bool BitReaderInterface::ReadVariableWidthS32(int32_t* val,
+                                              size_t chunk_length,
+                                              size_t zigzag_exponent) {
+  return ReadVariableWidthSigned(this, val, chunk_length, zigzag_exponent);
+}
+
+bool BitReaderInterface::ReadVariableWidthS16(int16_t* val,
+                                              size_t chunk_length,
+                                              size_t zigzag_exponent) {
+  return ReadVariableWidthSigned(this, val, chunk_length, zigzag_exponent);
+}
+
+bool BitReaderInterface::ReadVariableWidthS8(int8_t* val,
+                                             size_t chunk_length,
+                                             size_t zigzag_exponent) {
+  return ReadVariableWidthSigned(this, val, chunk_length, zigzag_exponent);
+}
+
+BitReaderWord64::BitReaderWord64(std::vector<uint64_t>&& buffer)
+    : buffer_(std::move(buffer)), pos_(0) {}
+
+BitReaderWord64::BitReaderWord64(const std::vector<uint8_t>& buffer)
+    : buffer_(ToBuffer64(buffer)), pos_(0) {}
+
+size_t BitReaderWord64::ReadBits(uint64_t* bits, size_t num_bits) {
+  assert(num_bits <= 64);
+  assert(IsLittleEndian() && "Big-endian architecture support not implemented");
+
+  if (ReachedEnd())
+    return 0;
+
+  // Index of the current word.
+  const size_t index = pos_ / 64;
+
+  // Bit position in the current word where we start reading.
+  const size_t offset = pos_ % 64;
+
+  // Read all bits from the current word (it might be too much, but
+  // excessive bits will be removed later).
+  *bits = buffer_[index] >> offset;
+
+  const size_t num_read_from_first_word = std::min(64 - offset, num_bits);
+  pos_ += num_read_from_first_word;
+
+  if (pos_ >= buffer_.size() * 64) {
+    // Reached end of buffer_.
+    return num_read_from_first_word;
+  }
+
+  if (offset + num_bits > 64) {
+    // Requested |num_bits| overflows to next word.
+    // Write all bits from the beginning of next word to *bits after offset.
+    *bits |= buffer_[index + 1] << (64 - offset);
+    pos_ += offset + num_bits - 64;
+  }
+
+  // We likely have written more bits than requested. Clear excessive bits.
+  *bits = GetLowerBits(*bits, num_bits);
+  return num_bits;
+}
+
+bool BitReaderWord64::ReachedEnd() const {
+  return pos_ >= buffer_.size() * 64;
+}
+
+bool BitReaderWord64::OnlyZeroesLeft() const {
+  if (ReachedEnd())
+    return true;
+
+  const size_t index = pos_ / 64;
+  if (index < buffer_.size() - 1)
+    return false;
```

Comment:
Could this return false even though there might only be zeroes left?     For example if there are a million zero bits left and nothing else, then we'll return false from here.  But the specification of the function says we should have returned true?  
If so, then the function definition should be clarified about this one-sided error.  Maybe it should be NoOnesLeft instead?